### PR TITLE
feat: allow card view when cardView setting is on

### PR DESCRIPTION
**Pull request by binodpant**
# Feature controlled card view vs list view toggling

The feature switch in config is off by default but when on, it allows switching between card and list views

Card view is default if feature is on, otherwise, table view is default (for now)

Current does not do the following:

- support usage within Data tAble's control bar, but it's fine for this UI since this one does not yet use Data Table bulk actions.
- use a Paragon component for toggle. It's a local toggle component used for now with a boolean switch to switch views, and also feature controlled with a flag
- finalize card look feel and behavior, that will be done in a separate ticket

# For all changes

- [x] Ensure adequate tests are in place (or reviewed existing tests cover changes)

# Only if submitting a visual change

- [x] Ensure to attach screenshots
- [ ] Ensure to have UX team confirm screenshots

https://user-images.githubusercontent.com/528166/142936365-1d58a333-68ba-43c8-80ec-d5b3587f7752.mov

This actually shows the buttons final look (I force hover style when you click on the iconbutton, for now). It gets us closer but not 100% compliance with Figma (we can't do 100% due to IconButton limitation)

https://user-images.githubusercontent.com/528166/142935649-17af8e9f-42e0-4cde-9647-94eaa0a82033.mov

### CODE REVIEW

File: src/components/catalogSearchResults/CatalogSearchResults.jsx [hunk 1]
```diff
@@ -6,12 +7,14 @@ import React, {
 import PropTypes from 'prop-types';
 import { connectStateResults } from 'react-instantsearch-dom';
 import {
-  Badge, DataTable, Alert, Button, useToggle,
+  Badge, DataTable, Alert, Button, useToggle, CardView, Card, Icon, IconButton,
```
Comment (kiram15) [inline src/components/catalogSearchResults/CatalogSearchResults.jsx]: nit: Organize these in alphabetical order and/or put them all on their own line for readability

File: src/components/catalogSearchResults/CatalogSearchResults.jsx [hunk 2]
```diff
@@ -109,6 +140,33 @@ export const BaseCatalogSearchResults = ({
...
+        onClick={() => { setCardView(true); }}
+        variant={cardView ? 'dark' : 'light'}
+        className="mr-2"
+      />
+      <IconButton
+        src={ListView}
+        iconAs={Icon}
+        alt="Card view"
+        onClick={() => { setCardView(false); }}
+        variant={!cardView ? 'dark' : 'light'}
```
Comment (kiram15) [inline src/components/catalogSearchResults/CatalogSearchResults.jsx]: The button should stay dark to show as selected (I think you're already working on this)

File: src/components/catalogSearchResults/CatalogSearchResults.jsx [hunk 3]
```diff
@@ -31,6 +34,34 @@ function formatDate(courseRun) {
...
+
+const CourseCard = ({ className, original }) => {
+  const { title, card_image_url, partners } = original;
+
+  return (
+    <Card className={className}>
+      <Card.Img
+        variant="top"
+        src={card_image_url}
+        style={{ width: '48vh', height: '26vh' }}
```
Comment (kiram15) [inline src/components/catalogSearchResults/CatalogSearchResults.jsx]: According to Figma, there should be 4 cards as a max when the screen can accommodate, right now I'm only seeing 3

File: .nvmrc [hunk 4]
```diff
@@ -1 +1 @@
-12.22.7
+v12
```
Comment (kiram15) [inline .nvmrc:1]: What is the purpose of the change in this file?

**Review by kiram15 — CHANGES_REQUESTED** [review]

File: src/components/catalogSearchResults/CatalogSearchResults.jsx [hunk 5]
```diff
@@ -109,6 +140,33 @@ export const BaseCatalogSearchResults = ({
...
   const [startDate, setStartDate] = useState();
   const [endDate, setEndDate] = useState();
+  const [cardView, setCardView] = useState(true);
+
+  // TODO: Feature control for Card view. Remove once cards are finalized
+  const config = getConfig();
+  const cardViewEnabled = config.FEATURE_CARD_VIEW_ENABLED === 'true';
+
+  // TODO: local view toggle compoent. To be replaced by IconButtonToggle from Paragon
+  const ViewToggle = () => (
```
Comment (kiram15) [inline src/components/catalogSearchResults/CatalogSearchResults.jsx]: We should probably also have tests for the toggle element to validate the change (which I also think you're working on atm)

File: .nvmrc (hunk 4 above)
Comment (binodpant) [reply .nvmrc:1]: ah this is just for future looking npm upgrade friendliness. Actually v12 will make the most recent 12 npm be used. But after my npm upgrade work is done this will instead change to v16. It does actually have no impact currently unless you run `nvm use` locally, then it will ask for installing the latest v12 node if not already

File: src/components/catalogSearchResults/CatalogSearchResults.jsx (hunk 5 above)
Comment (binodpant) [reply src/components/catalogSearchResults/CatalogSearchResults.jsx]: agreed! I am now working on some tests

File: src/components/catalogSearchResults/CatalogSearchResults.jsx (hunk 3 above)
Comment (binodpant) [reply src/components/catalogSearchResults/CatalogSearchResults.jsx]: ah good callout! I have to revisit how I did this, I will see if I can make 4 cards show per row instead. I don;t know if we had a min width requirement for these course cards though, mybe simply reducing maxWidth will work let's see

File: src/components/catalogSearchResults/CatalogSearchResults.jsx (hunk 2 above)
Comment (binodpant) [reply src/components/catalogSearchResults/CatalogSearchResults.jsx]: yup I managed to make it stay dark. Looks like I am missing making it actually color invert when selected like in the figma spec so will check that out

Comment (binodpant) [reply src/components/catalogSearchResults/CatalogSearchResults.jsx]: as discussed ,I got close, but not 100% per figma due to iconbutton limitation. we can revisit once we  work on the IconButton.Toggle component

**Review by kiram15 — APPROVED** [review]
LGTM! 👍 🚀

File: src/components/catalogSearchResults/CatalogSearchResults.jsx [hunk 6]
```diff
@@ -208,7 +296,18 @@ export const BaseCatalogSearchResults = ({
           pageSize={searchResults?.hitsPerPage || 0}
         >
           <DataTable.TableControlBar />
-          <DataTable.Table />
+          { cardViewEnabled && cardView ? (
+            <CardView
```
Comment (binodpant) [inline src/components/catalogSearchResults/CatalogSearchResults.jsx:300]: this uses the excellent column sizes api exposed by cardview to show requisite cols per row, given different screen sizes
